File: solutions/python/list_pivoting.py
from linked_list_prototype import ListNode


# @include
def list_pivoting(L, x):
    less_head = less_iter = ListNode()
    equal_head = equal_iter = ListNode()
    greater_head = greater_iter = ListNode()
    # Populates the three lists.
    while L:
        if L.data < x:
            less_iter.next = L
            less_iter = less_iter.next
        elif L.data == x:
            equal_iter.next = L
            equal_iter = equal_iter.next
        else:  # L.data > x.
            greater_iter.next = L
            greater_iter = greater_iter.next
        L = L.next
    # Combines the three lists.
    greater_iter.next = None
    equal_iter.next = greater_head.next
    less_iter.next = equal_head.next
    return less_head.next
# @exclude


def simple_test():
    L = ListNode(0)
    result = list_pivoting(L, 0)
    assert result is L
    result = list_pivoting(L, 1)
    assert result is L
    result = list_pivoting(L, -1)
    assert result is L

    L = ListNode(2, ListNode(0))
    result = list_pivoting(L, -1)
    assert result is L

    L = ListNode(2, ListNode(0))
    result = list_pivoting(L, 1)
    assert result.data == 0 and result.next.data == 2

    L = ListNode(2, ListNode(0, ListNode(-2)))
    result = list_pivoting(L, 1)
    assert result.data == 0 and result.next.data == -2 and result.next.next.data == 2


def main():
    simple_test()
    L = ListNode(1, ListNode(4, ListNode(3, ListNode(2, ListNode(5, None)))))
    x = 4
    result = list_pivoting(L, x)
    print()
    before_x = True
    while result:
        if result.data >= x:
            before_x = False
        if before_x:
            assert result.data < x
        else:
            assert result.data >= x
        print(result.data)
        result = result.next


if __name__ == '__main__':
    main()

File: solutions/cpp/list_pivoting.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>

#include "./Linked_list_prototype.h"
#include "./Merge_sorted_lists.h"  // uses AppendNode().

using std::cout;
using std::endl;
using std::make_shared;
using std::shared_ptr;

// @include
shared_ptr<ListNode<int>> ListPivoting(const shared_ptr<ListNode<int>>& L,
                                       int x) {
  shared_ptr<ListNode<int>> less_head(new ListNode<int>),
      equal_head(new ListNode<int>), greater_head(new ListNode<int>);
  shared_ptr<ListNode<int>> less_iter = less_head, equal_iter = equal_head,
                            greater_iter = greater_head;
  // Populates the three lists.
  shared_ptr<ListNode<int>> iter = L;
  while (iter) {
    AppendNode(&iter,
               iter->data < x
                   ? &less_iter
                   : iter->data == x ? &equal_iter : &greater_iter);
  }
  // Combines the three lists.
  greater_iter->next = nullptr;
  equal_iter->next = greater_head->next;
  less_iter->next = equal_head->next;
  return less_head->next;
}
// @exclude

void SimpleTest() {
  shared_ptr<ListNode<int>> L =
      make_shared<ListNode<int>>(ListNode<int>{0, nullptr});
  auto result = ListPivoting(L, 0);
  assert(result == L);
  result = ListPivoting(L, 1);
  assert(result == L);
  result = ListPivoting(L, -1);
  assert(result == L);

  L = make_shared<ListNode<int>>(ListNode<int>{
      2, make_shared<ListNode<int>>(ListNode<int>{0, nullptr})});
  result = ListPivoting(L, -1);
  assert(result == L);

  L = make_shared<ListNode<int>>(ListNode<int>{
      2, make_shared<ListNode<int>>(ListNode<int>{0, nullptr})});
  result = ListPivoting(L, 1);
  assert(result->data == 0 && result->next->data == 2);

  L = make_shared<ListNode<int>>(ListNode<int>{
      2, make_shared<ListNode<int>>(ListNode<int>{
             0, make_shared<ListNode<int>>(ListNode<int>{-2, nullptr})})});
  result = ListPivoting(L, 1);
  assert(result->data == 0 && result->next->data == -2 &&
         result->next->next->data == 2);
}

int main(int argc, char** argv) {
  SimpleTest();
  shared_ptr<ListNode<int>> L;
  L = make_shared<ListNode<int>>(ListNode<int>{
      1, make_shared<ListNode<int>>(ListNode<int>{
             4, make_shared<ListNode<int>>(ListNode<int>{
                    3, make_shared<ListNode<int>>(ListNode<int>{
                           2, make_shared<ListNode<int>>(
                                  ListNode<int>{5, nullptr})})})})});
  int x = 4;
  auto result = ListPivoting(L, x);
  cout << endl;
  bool before_x = true;
  while (result) {
    if (result->data >= x) {
      before_x = false;
    }
    if (before_x) {
      assert(result->data < x);
    } else {
      assert(result->data >= x);
    }
    cout << result->data << endl;
    result = result->next;
  }
  return 0;
}
